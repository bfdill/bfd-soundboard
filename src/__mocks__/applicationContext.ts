import { mockApplicationState } from './applicationState'
import { IApplicationContext } from '../types'
import { getJestLogger } from '../modules/winston-jest/index.test'

export const getMockApplicationContext = (): IApplicationContext =>
  ({
    state: mockApplicationState
  } as any)
