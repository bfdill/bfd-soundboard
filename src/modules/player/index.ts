// import * as playSound from 'play-sound'
import * as winston from 'winston'
import { ApplicationState } from '../../types'
import { createPlayer, IPlayer } from './types'
import { ISoundService, SoundService } from '../sounds'

// tsd creation fail
const createPlayer: createPlayer = require('play-sound')

export interface IPlayerService {
  playFile: (
    filename: string,
    state: ApplicationState,
    parentLogger: winston.Logger
  ) => PromiseLike<void>
}

export class PlayerService implements IPlayerService {
  static instance: IPlayerService | undefined

  constructor(
    protected readonly player: IPlayer,
    protected readonly soundService: ISoundService
  ) {}

  playFile = (
    filename: string,
    state: ApplicationState,
    parentLogger: winston.Logger
  ): PromiseLike<void> => {
    const logger = parentLogger.child({
      service: 'PlayerService',
      method: 'playFile'
    })

    logger.info({ filename })

    if (!this.soundService.isPathValid(filename, state, logger)) {
      const message = 'filename path is invalid'
      logger.error({ filename, message })
      return Promise.reject(message)
    }

    return new Promise((resolve, reject) => {
      const next = (error?: any) => {
        if (error === undefined || error === null) {
          logger.debug({
            filename,
            message: 'playback success'
          })
          resolve()
          return
        }

        logger.debug({
          error,
          filename,
          message: 'playback failure'
        })
        reject(error)
      }

      this.player.play(filename, next)
    })
  }

  static getInstance() {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerService(
      createPlayer(),
      SoundService.getInstance()
    )

    return this.instance
  }
}
