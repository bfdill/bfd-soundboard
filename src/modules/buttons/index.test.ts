import * as rpiGpio from 'rpi-gpio'
import { ButtonService, IButtonService, Button, GpioSetupConfig } from '.'
import { getMockSoundService } from '../sounds/__mocks__/soundService'
import { IPlayerService } from '../player'
import { getJestLogger } from '../winston-jest/index.test'

class TestButtonService extends ButtonService {
  getRegisteredSwitches = () => this.registeredSwitches
}

describe('modules -> buttons', () => {
  const mockOn: jest.Mock = jest.fn()
  const mockSetup: jest.Mock<void, []> = jest.fn()
  const gpio: rpiGpio.IGpio = {
    on: mockOn,
    promise: {
      setup: mockSetup
    }
  } as any
  const testButtons: Button[] = [
    { switch: '3', led: '8', soundId: '0' },
    { switch: '5', led: '10', soundId: '1' }
  ]
  const testSoundService = getMockSoundService()
  const testPlayerService: IPlayerService = {
    playFile: jest.fn()
  }
  const testButtonService = new TestButtonService(
    gpio,
    testButtons,
    testSoundService,
    testPlayerService
  )
  const buttonService: IButtonService = testButtonService
  const jestLogger = getJestLogger()

  test('has known exports', () => {
    expect(Object.keys(require('.'))).toMatchSnapshot()
  })

  afterEach(() => {
    jestLogger.callsMatchSnapshot()
  })

  describe('button service', () => {
    describe('startButtonService', () => {
      beforeEach(async () => {
        mockSetup.mockImplementation(() => Promise.resolve())

        await buttonService.startButtonService(jestLogger.logger)
      })

      test('setupFromConfig', () => {
        expect.hasAssertions()
        mockSetup.mock.calls.forEach(call => {
          expect(call).toMatchSnapshot()
        })
        jestLogger.callsMatchSnapshot()
      })

      test('registeredSwitches', () => {
        expect(testButtonService.getRegisteredSwitches()).toMatchSnapshot()
      })

      test('registerChangeListener', () => {
        expect(mockOn).toHaveBeenCalled()
      })

      test('error logs error', async () => {
        mockOn.mockImplementation(() => {
          throw new Error('borked')
        })
        const errorTest = new TestButtonService(
          gpio,
          testButtons,
          testSoundService,
          testPlayerService
        )
        const errorTestLogger = getJestLogger()

        await errorTest.startButtonService(errorTestLogger.logger)
        errorTestLogger.callsMatchSnapshot()
      })
    })

    describe('registerChangeListener', () => {
      let changeFn: Function
      beforeEach(async () => {
        mockSetup.mockImplementation(() => Promise.resolve())

        await buttonService.startButtonService(jestLogger.logger)

        changeFn = mockOn.mock.calls[0][1]
      })

      test('unknown channel(pin)', () => {
        changeFn('1', false)
        expect(testSoundService.getBySoundId).not.toHaveBeenCalled()
        expect(testPlayerService.playFile).not.toHaveBeenCalled()
      })

      test.each([false, true])('no action if active === pushed(%s)', state => {
        const sw = testButtons[0].switch
        const rs = testButtonService.getRegisteredSwitches()
        rs[sw].active = state

        changeFn(sw, !state)
        expect(testSoundService.getBySoundId).not.toHaveBeenCalled()
        expect(testPlayerService.playFile).not.toHaveBeenCalled()
      })

      test('not pushed = not active', () => {
        const sw = testButtons[0].switch
        const rs = testButtonService.getRegisteredSwitches()
        rs[sw].active = true

        changeFn(sw, true)
        expect(testSoundService.getBySoundId).not.toHaveBeenCalled()
        expect(testPlayerService.playFile).not.toHaveBeenCalled()
      })

      test('newly active getBySoundId', () => {
        expect.hasAssertions()
        testSoundService.setGetBySoundId(undefined)
        const sw = testButtons[0].switch
        const rs = testButtonService.getRegisteredSwitches()
        rs[sw].active = false

        try {
          changeFn(sw, false)
        } catch {}
        expect(testSoundService.getBySoundId).toHaveBeenCalled()
        expect(testPlayerService.playFile).not.toHaveBeenCalled()
      })

      test('newly active playFile', () => {
        testSoundService.setGetBySoundId({
          basename: 'test.mp3',
          filename: '/tmp/test.mp3',
          id: '0'
        })
        const sw = testButtons[0].switch
        const rs = testButtonService.getRegisteredSwitches()
        rs[sw].active = false

        changeFn(sw, false)
        expect(testSoundService.getBySoundId).toHaveBeenCalled()
        expect(testPlayerService.playFile).toHaveBeenCalled()
      })
    })

    describe('setupFromConfig', () => {
      beforeEach(async () => {
        mockSetup.mockImplementation(() => Promise.resolve())
        await buttonService.startButtonService(jestLogger.logger)
      })

      test('gpiop.setup', () => {
        mockSetup.mock.calls.forEach(call => {
          expect(call).toMatchSnapshot()
        })
      })

      test('config.cb', async () => {
        const config: GpioSetupConfig = {
          cb: jest.fn(),
          direction: 'out',
          edge: 'both',
          pin: '40'
        }
        await buttonService.setupFromConfig(config)
        expect(config.cb).toHaveBeenCalled()
      })
    })
  })

  describe('static members', () => {})
})
