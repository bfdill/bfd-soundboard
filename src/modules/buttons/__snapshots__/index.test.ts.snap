// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules -> buttons button service registerChangeListener newly active getBySoundId 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener newly active getBySoundId 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": true,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener newly active playFile 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener newly active playFile 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": true,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener no action if active === pushed(false) 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener no action if active === pushed(false) 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener no action if active === pushed(true) 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener no action if active === pushed(true) 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": true,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener not pushed = not active 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener not pushed = not active 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener unknown channel(pin) 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service registerChangeListener unknown channel(pin) 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service setupFromConfig config.cb 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service setupFromConfig config.cb 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service setupFromConfig gpiop.setup 1`] = `
Array [
  "8",
  "out",
  "none",
]
`;

exports[`modules -> buttons button service setupFromConfig gpiop.setup 2`] = `
Array [
  "3",
  "in",
  "both",
]
`;

exports[`modules -> buttons button service setupFromConfig gpiop.setup 3`] = `
Array [
  "10",
  "out",
  "none",
]
`;

exports[`modules -> buttons button service setupFromConfig gpiop.setup 4`] = `
Array [
  "5",
  "in",
  "both",
]
`;

exports[`modules -> buttons button service setupFromConfig gpiop.setup 5`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service setupFromConfig gpiop.setup 6`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService error logs error 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService error logs error 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService error logs error 3`] = `
Array [
  Object {
    "cause": [Error: borked],
    "level": "error",
    "message": "startButtonService error",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "error",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"cause\\":{},\\"message\\":\\"startButtonService error\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService error logs error 4`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService error logs error 5`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService registerChangeListener 1`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService registerChangeListener 2`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService registeredSwitches 1`] = `
Object {
  "3": Object {
    "active": false,
    "button": Object {
      "led": "8",
      "soundId": "0",
      "switch": "3",
    },
    "pushed": false,
  },
  "5": Object {
    "active": false,
    "button": Object {
      "led": "10",
      "soundId": "1",
      "switch": "5",
    },
    "pushed": false,
  },
}
`;

exports[`modules -> buttons button service startButtonService registeredSwitches 2`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService registeredSwitches 3`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 1`] = `
Array [
  "8",
  "out",
  "none",
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 2`] = `
Array [
  "3",
  "in",
  "both",
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 3`] = `
Array [
  "10",
  "out",
  "none",
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 4`] = `
Array [
  "5",
  "in",
  "both",
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 5`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 6`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 7`] = `
Array [
  Object {
    "level": "info",
    "message": "setup ðŸ‘œðŸ‘œðŸ‘œ complete",
    "method": "startButtonService",
    "service": "ButtonService",
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"setup ðŸ‘œðŸ‘œðŸ‘œ complete\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> buttons button service startButtonService setupFromConfig 8`] = `
Array [
  Object {
    "level": "debug",
    "message": "registeredSwitches",
    "method": "startButtonService",
    "registeredSwitches": Object {
      "3": Object {
        "active": false,
        "button": Object {
          "led": "8",
          "soundId": "0",
          "switch": "3",
        },
        "pushed": false,
      },
      "5": Object {
        "active": false,
        "button": Object {
          "led": "10",
          "soundId": "1",
          "switch": "5",
        },
        "pushed": false,
      },
    },
    "service": "ButtonService",
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":\\"ButtonService\\",\\"method\\":\\"startButtonService\\",\\"message\\":\\"registeredSwitches\\",\\"registeredSwitches\\":{\\"3\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"3\\",\\"led\\":\\"8\\",\\"soundId\\":\\"0\\"},\\"pushed\\":false},\\"5\\":{\\"active\\":false,\\"button\\":{\\"switch\\":\\"5\\",\\"led\\":\\"10\\",\\"soundId\\":\\"1\\"},\\"pushed\\":false}},\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> buttons has known exports 1`] = `
Array [
  "ButtonService",
]
`;
